src/filesystem_littlefs.c

```diff
@@ -0,0 +1,517 @@
+        return NULL;
+    }
+
+    path = process_path(path, devoptab_littlefs.name);
+
+    result = lfs_dir_open(lfs, dirState->dirStruct, path);
+    r->_errno = get_errno(result);
+
+    if (result < 0) {
+        return NULL;
+    }
+
+    return dirState;
+}
+
+static int dirreset_r(struct _reent *r, DIR_ITER *dirState) {
+    int result;
+
+    result = lfs_dir_rewind(lfs, dirState->dirStruct);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int dirnext_r(struct _reent *r, DIR_ITER *dirState, char *filename, struct stat *filestat) {
+    int result;
+    struct lfs_info info;
+
+    if (filename == NULL || filestat == NULL) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    result = lfs_dir_read(lfs, dirState->dirStruct, &info);
+    r->_errno = get_errno(result);
+
+    if (result == 0) {
+        // End of the directory
+        r->_errno = ENOENT; // has to be set at the end
+        return -1;
+    } else if (result < 0) {
+        return -1;
+    }
+
+    if (info.type & LFS_TYPE_DIR) {
+        filestat->st_mode = S_IFDIR;
+    } else if (info.type & LFS_TYPE_REG) {
+        filestat->st_mode = S_IFREG;
+    } else {
+        // Unexpected type, skip to the next dir
+        return dirnext_r(r, dirState, filename, filestat);
+    }
+
+    strncpy(filename, info.name, NAME_MAX);
+
+    return 0;
+}
+
+static int dirclose_r(struct _reent *r, DIR_ITER *dirState) {
+    int result;
+
+    result = lfs_dir_close(lfs, dirState->dirStruct);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int statvfs_r(struct _reent *r, const char *path, struct statvfs *buf) {
+    if (IS_EMPTY(path)) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    memset(buf, 0, sizeof(struct statvfs));
+
+    buf->f_frsize = lfs->cfg->block_size;
+    buf->f_bfree = lfs->free.size;
+    buf->f_bavail = buf->f_bfree;
+    buf->f_files = 0; // TODO: Count all inodes
+    buf->f_ffree = lfs->cfg->name_max - buf->f_files;
+    buf->f_favail = buf->f_ffree;
+    buf->f_fsid = (device & 0xFFFF) | crc32to16(lfs->seed); // 16b for filesystems, 16b for driver per filesystem
+    buf->f_flag = ST_NOSUID;
+    buf->f_namemax = lfs->cfg->name_max;
+
+    return 0;
+}
+
+static int ftruncate_r(struct _reent *r, void *fileStruct, off_t len) {
+    PREPARE_FIL_EX(f, fileStruct);
+    int result;
+
+    result = lfs_file_truncate(lfs, &(f->fil), len);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int rmdir_r(struct _reent *r, const char *path) {
+    return unlink_r(r, path);
+}
+
+static int lstat_r(struct _reent *r, const char *file, struct stat *st) {
+    // littlefs doesn't support links, just return stat
+    return stat_r(r, file, st);
+}
+
+static int utimes_r(struct _reent *r, const char *filename, const struct timeval times[2]) {
+    // Timestamps not implemented for littlefs
+    return 0;
+}
+
+static const devoptab_t devoptab_littlefs = {
+    .name = "internal",
```

Comment:
What is this name used for?

Reply:
It's used as a "mount point", you can access SPI flash under path `/internal/`.
Name chosen after discussion with @dragomirecky